pkg/controller/elasticsearch/pdb/reconcile.go

```diff
@@ -7,107 +7,165 @@ package pdb
 	"github.com/elastic/cloud-on-k8s/pkg/controller/common/defaults"
-	"github.com/elastic/cloud-on-k8s/pkg/controller/common/reconciler"
 	"github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/label"
 	"github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/name"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/sset"
 	"github.com/elastic/cloud-on-k8s/pkg/utils/k8s"
-	"k8s.io/api/policy/v1beta1"
-	"k8s.io/apimachinery/pkg/api/errors"
-	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	"k8s.io/apimachinery/pkg/runtime"
 )
 
-// Reconcile ensures that a PodDisruptionBudget exists for this cluster according to the spec.
-//
-// If the spec has disabled the default PDB, it will ensure it does not exist
-func Reconcile(
-	c k8s.Client,
-	scheme *runtime.Scheme,
-	es v1alpha1.Elasticsearch,
-) error {
-	disabled := false
-
-	template := es.Spec.PodDisruptionBudget
-	if template != nil {
-		// clone to avoid accidentally overwriting template fields
-		template = template.DeepCopy()
-
-		emptyTemplate := commonv1alpha1.PodDisruptionBudgetTemplate{}
-		if reflect.DeepEqual(&emptyTemplate, template) {
-			disabled = true
-		}
-	} else {
-		template = &commonv1alpha1.PodDisruptionBudgetTemplate{}
+// Reconcile ensures that a PodDisruptionBudget exists for this cluster, inheriting the spec content.
+// The default PDB we setup dynamically adapts MinAvailable to the number of nodes in the cluster.
+// If the spec has disabled the default PDB, it will ensure none exist.
+func Reconcile(k8sClient k8s.Client, scheme *runtime.Scheme, es v1alpha1.Elasticsearch, statefulSets sset.StatefulSetList) error {
+	expected, err := expectedPDB(es, statefulSets, scheme)
+	if err != nil {
+		return err
 	}
-
-	var objectMeta metav1.ObjectMeta
-	if template != nil {
-		objectMeta = *template.ObjectMeta.DeepCopy()
+	if expected == nil {
+		return deleteDefaultPDB(k8sClient, es)
 	}
-	objectMeta.Name = name.DefaultPodDisruptionBudget(es.Name)
-	objectMeta.Namespace = es.Namespace
-	objectMeta.Labels = defaults.SetDefaultLabels(objectMeta.Labels, label.NewLabels(k8s.ExtractNamespacedName(&es)))
 
-	expected := v1beta1.PodDisruptionBudget{
-		ObjectMeta: objectMeta,
-		Spec:       template.Spec,
+	// reconcile actual vs. expected
+	var actual v1beta1.PodDisruptionBudget
+	err = k8sClient.Get(k8s.ExtractNamespacedName(expected), &actual)
+	if err != nil && apierrors.IsNotFound(err) {
+		return k8sClient.Create(expected)
 	}
 
-	if disabled {
-		// delete the default budget if it exists.
-		//
-		// we do this by getting first because that is a local cache read,
-		// versus a Delete call, which would hit the API.
-
-		if err := c.Get(k8s.ExtractNamespacedName(&expected), &expected); err != nil && !errors.IsNotFound(err) {
-			return err
-		} else if errors.IsNotFound(err) {
-			// already deleted, which is fine
-			return nil
+	// update if we're missing a label
+	needsLabelUpdate := false
+	for k, v := range expected.Labels {
```

Comment:
What if a label was removed? Also, should this be made a helper somewhere?

Reply:
The problem is: how do you know a label should be removed ? Maybe you see one that should not be there but that's because it was added automatically by a mutation webhook.

I guess the correct way to handle that is to rely on the hash label we use in other places. It sounded a bit overkill to me, but maybe better to use it here as well (also, it simplifies comparison).

Reply:
Last commit switches to hash-based comparison 👍

Reply:
Nice.